vault/raft.go

```diff
@@ -573,23 +602,73 @@ func (c *Core) InitiateRetryJoin(ctx context.Context) error {
 }
 
 func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJoinInfo, nonVoter bool) (bool, error) {
-	raftStorage, ok := c.underlyingPhysical.(*raft.RaftBackend)
-	if !ok {
-		return false, errors.New("raft storage not configured")
+	raftBackend := c.getRaftBackend()
+	if raftBackend == nil {
+		return false, errors.New("raft backend not in use")
 	}
 
-	if raftStorage.Initialized() {
-		return false, errors.New("raft storage is already initialized")
-	}
+	// NOTE: We need to support initialized core for the ha_storage case
 
 	init, err := c.Initialized(ctx)
 	if err != nil {
 		return false, errwrap.Wrapf("failed to check if core is initialized: {{err}}", err)
 	}
-	if init {
+
+	_, isRaftStorage := c.underlyingPhysical.(*raft.RaftBackend)
+
+	// Prevent join from happening if we're using raft for storage and
+	// it has already been initialized.
+	if init && isRaftStorage {
 		return true, nil
 	}
 
+	if isRaftStorage && !c.Sealed() {
+		c.logger.Error("node must be sealed before joining")
+		return false, errors.New("node must be sealed before joining")
+	}
+
+	isRaftHA := c.isRaftHAOnly()
+	// Disallow leader API address to be provided if we're using raft for HA-only
+	if isRaftHA {
+		for _, info := range leaderInfos {
+			if info.LeaderAPIAddr != "" {
+				return false, errors.New("leader API address must be unset when raft is used exclusively for HA")
+			}
+		}
+
+		// Get the leader address from storage
+		keys, err := c.barrier.List(ctx, coreLeaderPrefix)
+		if err != nil {
+			return false, err
+		}
+
+		if len(keys) == 0 || len(keys[0]) == 0 {
+			return false, errors.New("unable to fetch leadership entry")
+		}
+
+		uuid := coreLeaderPrefix + keys[0]
+		entry, err := c.barrier.Get(ctx, uuid)
+		if err != nil {
+			return false, err
+		}
+		if entry == nil {
+			return false, errors.New("unable to read leadership entry")
+		}
+
+		var adv activeAdvertisement
+		err = jsonutil.DecodeJSON(entry.Value, &adv)
+		if err != nil {
+			return false, errwrap.Wrapf("unable to decoded leader entry: {{err}}", err)
+		}
+
+		leaderInfos[0].LeaderAPIAddr = adv.RedirectAddr
+	}
+
+	if !isRaftHA && !c.Sealed() {
+		c.logger.Error("node must be sealed before joining")
+		return false, errors.New("node must be sealed before joining")
```

Comment:
```suggestion
		return false, errors.New("node must be unsealed before joining")
```

Reply:
I committed the suggestion above too quick -- this is for the case where raft is used for storage, so it needs to be sealed.